File: Arcade/Intro/12.clj
(defn insertByHeight [lst_ lst_pos]
  "insertByheight : [Int], [Int] -> [Int].
  Insert lst_pos without moving -1 elements."
  (cond
    (empty? lst_) nil
    (pos? (first lst_)) (cons (first lst_pos) (insertByHeight (rest lst_) (rest lst_pos)))
    :else (cons (first lst_) (insertByHeight (rest lst_) lst_pos))))

(defn sortByHeight [a]
  "sortbyheight : [Int] -> ([Int], [Int] -> [Int]) -> [Int].
  Sort a without sorting -1."
  (let [lst (sort (filter pos? a))]
    (insertByHeight a lst)))

File: Arcade/Intro/12.cpp
#include<vector>

std::vector<int> sortByHeight(std::vector<int> a) {
  // Sort a without sorting -1.
  for(int i = 0; i < a.size(); i++) { // adapt-inserctiosort
    if(a[i] == -1) continue;
    for(int j = i + 1; j < a.size(); j++) {
      if(a[j] == -1) continue;
      if(a[j] < a[i]) std::swap(a[i], a[j]);
    }
  }
  return a;
}
